### PR TITLE
SL-18330: Support parsing direct from stream; improve C++ compatibility.

### PR DESCRIPTION
**Pull request by nat-goodspeed**
The major feature introduced by this PR is that `llsd.parse()` can now accept a seekable binary stream instead of requiring the caller to read the stream (e.g. a file) into memory before parsing.

It also improves recognition of MIME-type headers produced by the C++ LLSD library embedded in the Second Life viewer.

### CODE REVIEW

File: llsd/base.py [hunk 1]
```diff
@@ -483,19 +477,197 @@ def _parse_string_delim(self, delim):
...
+        return self._buffer[self._index:self._index + num]
+
+    def getc(self, num=1, full=True):
+        # full=True means error if we can't read num bytes
+        chars = self.peek_impl(num, full)
+        self._index += num
+        return chars
+
+
+class StreamUtil:
```
Comment (bennettgoble) [inline llsd/base.py:517]: Instead of building wrappers around bytes and streams, could `llsd.parse` convert all input into a unified class such as `BufferedReader` which provides the requisite peek implementation? This would make these classes unneeded.

File: llsd/serde_binary.py [hunk 2]
```diff
@@ -232,9 +231,22 @@ def parse_binary(something):
...
+    # If we matched the header, then parse whatever follows, else parse the
+    # original bytes object or stream.
+    return parse_binary_noh(remainder if remainder is not None else something)
+
+
+def match_binary_hdr(something):
+    return matchseq(something, b'<? llsd/binary ?>')
 
+
+def parse_binary_noh(something):
```
Comment (bennettgoble) [inline llsd/serde_binary.py:245]: `"noh"` is cute but its meaning is not obvious. Could the header option be an optional kwarg of the simpler `parse_*` functions or could the name be made clearer?

File: llsd/base.py (hunk 1 above)
Comment (brad-linden) [reply llsd/base.py:517]: +1 interested to hear discussion about this as well

**Review by brad-linden — APPROVED** [review]
conditional on resolving the question about BufferedReader usage above, this PR looks good to me

**Comment by brad-linden** [comment]
also, per recent [developments](https://lindenlab.atlassian.net/wiki/spaces/dx/blog/2023/02/10/2829222196/Basics+image+update+Jessie+is+deprecated+no+longer+built), I think it is not necessarily worth maintaining python2 and 3.4 fallback code for new enhancement work like this.  If one wanted to juice our codecov metrics, it could be done by ripping some of that old stuff out.

File: llsd/serde_binary.py (hunk 2 above)
Comment (nat-goodspeed) [reply llsd/serde_binary.py:245]: How about `parse_binary_nohdr()`? Or even `_parse_binary_nohdr()` since I view this as primarily internal?

I want to be able to pass control directly to this code from `llsd.parse()` when the generic parser recognizes a particular header, instead of passing and testing a bool.

File: llsd/base.py (hunk 1 above)
Comment (nat-goodspeed) [reply llsd/base.py:517]: The distinction between `StreamUtil` and `BytesUtil` becomes moot when `LLSDBaseParser.force_stream` is `True`, which it is by default: in that case, instead of using `BytesUtil` at all, `_reset()` copies the input `bytes` object into an `io.BytesIO` and proceeds with `StreamUtil`.

I was curious whether a `BytesUtil` wrapper would have any performance benefits over the tactic above. I suspect it only matters if the consumer passes a very large `bytes` object, in which case both the buffer allocation and the copy itself would incur noticeable overhead. But - in the likely case that the consumer is reading that very large `bytes` object from a stream, we could address that overhead by passing the stream directly to `parse()`.

Bottom line: we can probably make `BytesUtil` go away entirely. In that case, yes, we could wrap the input stream (whether passed by the caller or an internal `BytesIO`) in a `BufferedReader` and use its `peek()`. It might still be useful to wrap it in a class that provides `StreamUtil.error()`.

Is that the direction we should go?

Comment (nat-goodspeed) [reply llsd/base.py:517]: `BytesStream` and `AnyStream` could similarly be elided.

Comment (bennettgoble) [reply llsd/base.py:517]: I'd like to see SteamUtil, MatchStream, AnyStream, BytesStream disappeared, they're confusing and, imo, a too complicated when we can just take input and coerce it into a standard I/O contract.

File: llsd/base.py [hunk 3]
```diff
@@ -375,36 +370,41 @@ class LLSDBaseParser(object):
...
+            util = StreamUtil(something)
+        else:
+            raise LLSDParseError('Input must either be a bytes object '
+                                 'or a seekable binary stream')
+        # Instead of indirecting through 'util' on every access, directly
+        # capture its published methods as methods on this instance, but
+        # prepend an underscore to each method name.
+        for attr in dir(util):
+            if not attr.startswith('_'):
+                setattr(self, '_' + attr, getattr(util, attr))
```
Comment (bennettgoble) [inline llsd/base.py:397]: Why...? Let's avoid metaprogramming for minor convenience reasons if possible.

Comment (nat-goodspeed) [reply llsd/base.py:397]: That was actually intended for performance: avoiding an extra hash lookup. There's code in our parsers that uses the Python idiom of capturing certain object attributes as local variables. I don't often do that myself because I'm not often dealing with performance-critical Python code, and I prefer clarity. That said, the lines above are a nod to the theoretical performance benefit of skipping an extra indirection.

But with the local stream wrappers vanishing, the above becomes moot and will likewise disappear.

**Comment by nat-goodspeed** [comment]
Since this PR has become essential for ongoing DRTVWR-558 Puppetry work, absent further objections or requested changes, I'm going to merge this soonish.
